src/main/java/build/buildfarm/worker/ReportResultStage.java

```diff
@@ -186,19 +185,112 @@ protected OperationContext tick(OperationContext operationContext) throws Interr
           .setIsExecutable(Files.isExecutable(outputPath));
       inlineContentBytes = inlineOrDigest(
           content,
-          policy,
+          workerContext.getFileCasPolicy(),
           contents,
           inlineContentBytes,
           workerContext.getInlineContentLimit(),
           () -> outputFileBuilder.setContent(content),
           (fileContent) -> outputFileBuilder.setDigest(getDigestUtil().compute(fileContent)));
     }
 
+    for (String outputDir : outputDirs) {
+      Path outputDirPath = root.resolve(outputDir);
+      if (!Files.exists(outputDirPath)) {
+        continue;
+      }
+
+      Tree.Builder treeBuilder = Tree.newBuilder();
+      Directory.Builder outputRoot = treeBuilder.getRootBuilder();
+      Files.walkFileTree(outputDirPath, new SimpleFileVisitor<Path>() {
+        Directory.Builder currentDirectory = null;
+        Stack<Directory.Builder> path = new Stack<>();
+
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+          ByteString content;
+          try (InputStream inputStream = Files.newInputStream(file)) {
+            content = ByteString.readFrom(inputStream);
+          } catch (IOException e) {
+            e.printStackTrace();
+            return FileVisitResult.CONTINUE;
+          }
+
+          // should we cast to PosixFilePermissions and do gymnastics there for executable?
+
+          // TODO symlink per revision proposal
+          contents.add(content);
+          FileNode.Builder fileNodeBuilder = currentDirectory.addFilesBuilder()
+              .setName(file.getFileName().toString())
+              .setDigest(getDigestUtil().compute(content))
+              .setIsExecutable(Files.isExecutable(file));
+          return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+          path.push(currentDirectory);
```

Comment:
Are you sure you need to push current directory when it is equal to the root? It seems to break the invariant that currentDirectory's parent is equal to stack.top